python/xoscar/collective/core.py

```diff
@@ -0,0 +1,321 @@
+# Copyright 2022-2023 XProbe Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import hashlib
+import os
+from collections import defaultdict
+from typing import Any, Dict, List, Optional
+
+from .. import Actor, actor_ref
+from ..context import get_context
+from .common import (
+    INVOKE_ERROR_MESSAGE,
+    RANK_ADDRESS_ENV_KEY,
+    AllReduceAlgorithm,
+    CollectiveReduceOp,
+)
+from .process_group import ProcessGroup, ProcessGroupGloo
+from .utils import get_rank_address_via_env
+
+
+class RankActor(Actor):
+    def __init__(
+        self,
+        rank: int,
+        world: int,
+        backend: str = "gloo",
+        pg_options: Optional[ProcessGroup.Options] = None,
+        *args,
+        **kwargs,
+    ):
+        self._rank = rank
+        self._world = world
+        self._backend = backend
+        self.name_to_pg: Dict[str, Dict[str, "ProcessGroup"]] = defaultdict(dict)
+        self._pg_options = pg_options
+
+    @classmethod
+    def default_uid(cls):
+        return "RankActor"
+
+    async def __post_create__(self):
+        os.environ[RANK_ADDRESS_ENV_KEY] = self.address
+        _ip = self._get_ip()
+        if self._backend == "gloo":
+            pg = ProcessGroupGloo(
+                _ip,
+                self._rank,
+                self._world,
+                group_name="default",
+                pg_options=self._pg_options,
+            )
+            self.name_to_pg["gloo"]["default"] = pg
```

Comment:
What happend if the `self._backend` != "gloo"?

Reply:
> What happend if the `self._backend` != "gloo"?

Here to leave a ``else`` to ``nccl`` impl. I can add a raise here for now.